### DISCUSSION TITLE
What are the OS dependent portions?

### DISCUSSION THREAD

**Discussion by lonnietc** [cloud-hypervisor/cloud-hypervisor#2863] [Q&A]
Hello,

I have a hypervisor project and am looking for a modern VMM to have running in the VM's.

There are a number of existing possibilities like porting QEMU, Bhyve, or VirtualBox, to name a few, all of which have a strong industry presence and are also C/C++ but all present significant challenges.

When I came across Cloud-Hypervisor by finding Firecracker and the Rust-VMM project, it made me think that perhaps a "modern" and lightweight VMM might be the way to go as well as the fact that I truly like the ideas embodied within Cloud-Hypervisor for having a component based design which is simply awesome.

Of course, my background is in C/C++ and other similar languages so I am going to spend some time to get up to speed on Rust which I also find has some very attractive features over C/C++ as well which is another driver for this possible effort.

With all of this in mind and knowing that Cloud-Hypervisor seems to have the only OS dependent parts for KVM or MSHV , I am wondering how difficult it might be to port the VMM over to a new hypervisor platform which is actually closer to bare metal and an OS?

Any ideas or thoughts on this would be greatly appreciated.

Cheers and have a great day,
Lonnie

**Comment by sboeuf** [comment]
Well it's good to see your interest in the Cloud Hypervisor project :)

About porting it to a different hypervisor, it is fairly straightforward if you follow what has been done in the `hypervisor` crate to support both KVM and MSHV. But of course that's assuming you're gonna run it on top of Linux/Unix. Otherwise you would have to adapt all the places with dedicated system calls so that they can work with the underlying OS as well.
So to summarize, the complexity of this work will depend on the underlying interfaces Cloud Hypervisor is going to interact with.

**Reply by lonnietc** [reply]
Yea, I think that the Cloud Hypervisor shows great promise although I also think that it will take a lot of work, which I do not mind doing if I can get a feel for Rust as well as what areas are making system calls.

The challenge is that the new hypervisor was written in C/C++ and has no standard libraries and I was currently looking to port the MUSL Clib over to it for use with the small existing framework that it has now but started to consider if using Rust and the Cloud Hypervisor VMM might be a better solution than porting over one of the existing hypervisors. 

I'll have to keep digging into the code to see what is the best approach to get it done.
